crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+        // last finalized block would be number 9.
+        let mut tree = BlockchainTree::new(externals, 1, 2, 3).unwrap();
+
+        // genesis block 10 is already canonical
+        assert_eq!(tree.make_canonical(&H256::zero()), Ok(()));
+
+        // insert block2 hits max chain size
+        assert_eq!(
+            tree.insert_block_with_senders(&block2),
+            Err(ExecError::PendingBlockIsInFuture {
+                block_number: block2.number,
+                block_hash: block2.hash(),
+                last_finalized: 9,
+            }
+            .into())
+        );
+
+        // make genesis block 10 as finalized
+        tree.finalize_block(10);
+
+        // block 2 parent is not known.
+        assert_eq!(tree.insert_block_with_senders(&block2), Ok(false));
+
+        // insert block1
+        assert_eq!(tree.insert_block_with_senders(&block1), Ok(true));
+        // already inserted block will return true.
+        assert_eq!(tree.insert_block_with_senders(&block1), Ok(true));
+
+        // insert block2
+        assert_eq!(tree.insert_block_with_senders(&block2), Ok(true));
+
+        // Trie state:
+        //      b2 (pending block)
+        //      |
+        //      |
+        //      b1 (pending block)
+        //    /
+        //  /
+        // g1 (canonical blocks)
+        // |
+
+        // make block1 canonical
+        assert_eq!(tree.make_canonical(&block1.hash()), Ok(()));
+        // make block2 canonical
+        assert_eq!(tree.make_canonical(&block2.hash()), Ok(()));
+
+        // Trie state:
+        // b2 (canonical block)
+        // |
+        // |
+        // b1 (canonical block)
+        // |
+        // |
+        // g1 (canonical blocks)
+        // |
+
+        let mut block1a = block1.clone();
+        let block1a_hash = H256([0x33; 32]);
+        block1a.block.header.hash = block1a_hash;
+        let mut block2a = block2.clone();
+        let block2a_hash = H256([0x34; 32]);
+        block2a.block.header.hash = block2a_hash;
+
+        // reinsert two blocks that point to canonical chain
+        assert_eq!(tree.insert_block_with_senders(&block1a), Ok(true));
+        assert_eq!(tree.insert_block_with_senders(&block2a), Ok(true));
+
+        // Trie state:
+        // b2   b2a (side chain)
+        // |   /
+        // | /
+        // b1  b1a (side chain)
+        // |  /
+        // |/
+        // g1 (10)
+        // |
+        assert_eq!(tree.chains.len(), 2);
+        assert_eq!(
+            tree.block_indices.blocks_to_chain,
+            HashMap::from([(block1a_hash, 1), (block2a_hash, 2)])
+        );
+        assert_eq!(
+            tree.block_indices.fork_to_child,
+            HashMap::from([
+                (block1.parent_hash, HashSet::from([block1a_hash])),
+                (block1.hash(), HashSet::from([block2a_hash]))
+            ])
+        );
+
+        // make b2a canonical
+        assert_eq!(tree.make_canonical(&block2a_hash), Ok(()));
+        // Trie state:
```

Comment:
can we add assertions like you did above that indeed the tree state is correct?

Reply:
Added